cidc_api/models/models.py

```diff
@@ -488,9 +523,37 @@ def insert(self, session: Session, commit: bool = True, compute_etag: bool = Tru
             )
 
         logger.info(
-            f"admin-action: {grantor.email} gave {grantee.email} the permission {self.upload_type} on {self.trial_id}"
+            f"admin-action: {grantor.email} gave {grantee.email} the permission {self.upload_type or 'all assays'} on {self.trial_id or 'all trials'}"
+        )
+
+        # If this is a permission granting the user access to all trials for
+        # a given upload type or all upload types for a given trial, delete
+        # any related trial-upload type specific permissions to avoid
+        # redundancy in the database and in conditional IAM bindings.
+        perms_to_delete = (
+            session.query(Permissions)
+            .filter(
+                Permissions.granted_to_user == self.granted_to_user,
+                # If inserting a cross-trial perm, then select relevant
+                # trial-specific perms for deletion.
+                Permissions.trial_id != self.EVERY
+                if self.trial_id == self.EVERY
+                else Permissions.trial_id == self.trial_id,
+                # If inserting a cross-upload type perm, then select relevant
+                # upload type-specific perms for deletion.
+                Permissions.upload_type != self.EVERY
+                if self.upload_type == self.EVERY
+                else Permissions.upload_type == self.upload_type,
+            )
+            .all()
         )
 
+        # Add any related permission deletions to the insertion transaction.
+        # If a delete operation fails, all other deletes and the insertion will
+        # be rolled back.
+        for perm in perms_to_delete:
+            perm.delete(deleted_by=self.granted_by_user, commit=False)
```

Comment:
```suggestion
            perm.delete(deleted_by=self.granted_by_user, session=session, commit=False)
```

Reply:
Since we loaded `perms_to_delete` via `session.query`, I think each instance is already bound to `session`. I'll double check this, though.

Reply:
well for the readability sake i'd still add it

Reply:
took another, similar route (`session.delete(perm)`) - the readability argument is a good one!